DynamicStoragePiles/RestrictContainers.cs

```diff
@@ -0,0 +1,298 @@
+        ///     removal of items from the source inventory can be prevented if they were not added to the container.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="fromInventory"></param>
+        [HarmonyPrefix]
+        [HarmonyPatch(typeof(Inventory), nameof(Inventory.MoveAll), new[] { typeof(Inventory) })]
+        [HarmonyPriority(Priority.First)]
+        private static void MoveAllPrefix(Inventory __instance, Inventory fromInventory) {
+            if (__instance == null || fromInventory == null) { return; }
+            
+            Jotunn.Logger.LogDebug("MoveAllPrefix");
+            Jotunn.Logger.LogDebug($"Move to: {__instance.m_name}");
+
+            if (IsRestrictedContainer(__instance.m_name, out string allowedItem)) {
+                _targetContainer = __instance.m_name;
+                _allowedItem = allowedItem;
+            }
+        }
+
+        /// <summary>
+        ///     Reset the _targetContainer and _allowedItem values since MoveAll has finished.
+        /// </summary>
+        [HarmonyPrefix]
+        [HarmonyPatch(typeof(Inventory), nameof(Inventory.MoveAll), new[] { typeof(Inventory) })]
+        [HarmonyPriority(Priority.First)]
+        private static void MoveAllPostfix() {
+            _targetContainer = null;
+            _allowedItem = null;
+        }
+
+        /// <summary>
+        ///     Patch to check trigger ShouldRemoveItem check and prevent RemoveItem from running if
+        ///     the item was moved during a call to MoveAll and failed to be added to the target
+        ///     container because it was restricted and not the allowed item type.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(typeof(Inventory), nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool RemoveItemPrefix_1(Inventory __instance, ItemDrop.ItemData item) {
+            return ShouldRemoveItem(__instance, item);
+        }
+
+        /// <summary>
+        ///     Patch to check trigger ShouldRemoveItem check and prevent RemoveItem from running if
+        ///     the item was moved during a call to MoveAll and failed to be added to the target
+        ///     container because it was restricted and not the allowed item type.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData), typeof(int) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool RemoveItemPrefix_2(Inventory __instance, ItemDrop.ItemData item) {
+            return ShouldRemoveItem(__instance, item);
+        }
```

Comment:
```suggestion
        /// <summary>
        ///     Patch to check trigger ShouldRemoveItem check and prevent RemoveItem from running if
        ///     the item was moved during a call to MoveAll and failed to be added to the target
        ///     container because it was restricted and not the allowed item type.
        /// </summary>
        /// <param name="__instance"></param>
        /// <param name="item"></param>
        /// <returns></returns>
        [HarmonyPrefix]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData) })]
        [HarmonyPatch(typeof(Inventory), nameof(Inventory.RemoveItem), new[] { typeof(ItemDrop.ItemData), typeof(int) })]
        [HarmonyPriority(Priority.First)]
        private static bool RemoveItemPrefix(Inventory __instance, ItemDrop.ItemData item) {
            return ShouldRemoveItem(__instance, item);
        }
```

Here the two patches are also the same. Also, `typeof(Inventory)` was missing here as well